<a href="https://colab.research.google.com/github/egitonga395/DRFAUTHSYSTEM/blob/main/Erick_data_manipulation_with_python_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>


# Question 1
What is the primary purpose of data manipulation in the data science workflow?
1. Helps with dealing with duplicate data
2. Missing data
3. Removing duplicates
4. Correcting the data type fomarts

# Question 2
Name three Python libraries that are used for data manipulation, besides Pandas.
1. Pyspark
2. polar
3. Dask

# Question 3
What are the key limitations of standard Python lists when performing numerical computations with large datasets, which NumPy addresses?
1. Numpy is very fast as it with numerical data while compared to lists
2. Numpy's syntax is allowed for mathematical operations without the need of for loops which adds on to the complexity and utiliztion of more computing resources

# Question 4
Explain the core difference between how Python lists and NumPy arrays handle the addition of two lists/arrays.
1. While list concantete when mathematical operations are introcude, Numpy allows for direct mathematical operations
2. It is difficult to visualize the arrays in list as it involves nesting  of 2 lists numpy on the other hands presents the data as columns and rows

# Question 5
What is an ndarray

An Ndarray refers to Nth dimensional Array, This refers to data dat which is orgganized in an array, which could be nested in each other. This data is often used to give representation of real world models, for example distance is a one dimension, speed is two dimension[matrix], volume is three dimension, space-time as 4d  

#Question 6
List the advantages of using NumPy arrays over Python lists for numerical computations.
1. Numpy is faster than list interms of computation
2. Numpy is much efficient interms of computer resources
3. Numpy is much easier to use
4. Numpy is much more intuitive, especially when use with pandas model the data visually

# Question 7
What are the two primary data structures that Pandas introduces for working with tabular data? Briefly describe each.
1. Series is used with one dimensional data
2. Dataframe is used to deal with two dimensional data such as spreadsheets and datasets/tables from databases

#Question 8
How does Pandas simplify working with tabular data compared to what NumPy primarily offers?
Pandas arranges the data into rows and columns making the data easier to use
futhermore the data can be indexed to ensure that it is intuitive

#Question 9
Name two ways to access a specific column in a Pandas DataFrame. Provide a simple example for each.
df["column_name"]
df.column_name
supposed your dataframe has a colums called order_id that would like to be accessed
df["order_id"]
or
df.order_id
can be used
# Question 10
What does df.isnull().sum() tell you about a DataFrame in Pandas, and why is this function important in data cleaning?

df.isnull().sum() is used to check for the total NaN values in a dataset being analysed. It is important to so so as to find out whether some row values will be dropped or repopulated. This is done inorder to avoid errors during mathematical computations. Null values often cause errors during analysis since they are the absence of a value.


#### Section 2: Foundational Data Exploration & Manipulation (10 Questions)


Loading the sheets

In [46]:
import pandas as pd

df = pd.read_csv("https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv", sep='\t')

1.  Display the last 7 rows of the DataFrame `df` to get an initial look at the data.


Look at the data first


In [47]:
df.tail(7)

,order_id,quantity,item_name,choice_description,item_price
4615,1832,1,Chicken Soft Tacos,"[Fresh Tomato Salsa, [Rice, Cheese, Sour Cream]]",$8.75
4616,1832,1,Chips and Guacamole,NaN,$4.45
4617,1833,1,Steak Burrito,"[Fresh Tomato Salsa, [Rice, Black Beans, Sour ...",$11.75
4618,1833,1,Steak Burrito,"[Fresh Tomato Salsa, [Rice, Sour Cream, Cheese...",$11.75
4619,1834,1,Chicken Salad Bowl,"[Fresh Tomato Salsa, [Fajita Vegetables, Pinto...",$11.25
4620,1834,1,Chicken Salad Bowl,"[Fresh Tomato Salsa, [Fajita Vegetables, Lettu...",$8.75
4621,1834,1,Chicken Salad Bowl,"[Fresh Tomato Salsa, [Fajita Vegetables, Pinto...",$8.75


2.  Print a summary of the DataFrame including the index dtype and column dtypes, non-null values, and memory usage. Comment on what you see.


Look at the description of the data

In [48]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4622 entries, 0 to 4621
Data columns (total 5 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   order_id            4622 non-null   int64 
 1   quantity            4622 non-null   int64 
 2   item_name           4622 non-null   object
 3   choice_description  3376 non-null   object
 4   item_price          4622 non-null   object
dtypes: int64(2), object(3)
memory usage: 180.7+ KB


There are 4622 counts.  Choice descrition has NaNs.

3.  Determine how many observations have missing item|_price in the dataset.


How many entries item_price is none in the dataset

In [49]:
df["item_price"].isnull().sum()

np.int64(0)

4.  Print the names of all columns in the DataFrame.


What are the columns names in the dataset

In [50]:
print(df.columns)

Index(['order_id', 'quantity', 'item_name', 'choice_description',
       'item_price'],
      dtype='object')


5.  Identify which `item_name` appears most frequently in the orders.


Identify the item name that appears most frequestly

In [51]:

df.item_name.mode()

,item_name
0,Chicken Bowl


6.  Determine how many distinct `item_name` values are present in the dataset.


how many distict entries are here? non-duplicated data. Item name does not contaun any NaN values

In [52]:
print(df.item_name.count() - df.duplicated().sum())

4563


7.  Create a new DataFrame containing only the rows where the `item_name` is 'Chicken Bowl'.


creating a new dataframe which only has the name chickenbowl

In [53]:
df2 = df[df['item_name'] == 'Chicken Bowl']
df2.head()

,order_id,quantity,item_name,choice_description,item_price
4,2,2,Chicken Bowl,"[Tomatillo-Red Chili Salsa (Hot), [Black Beans...",$16.98
5,3,1,Chicken Bowl,"[Fresh Tomato Salsa (Mild), [Rice, Cheese, Sou...",$10.98
13,7,1,Chicken Bowl,"[Fresh Tomato Salsa, [Fajita Vegetables, Rice,...",$11.25
19,10,1,Chicken Bowl,"[Tomatillo Red Chili Salsa, [Fajita Vegetables...",$8.75
26,13,1,Chicken Bowl,"[Roasted Chili Corn Salsa (Medium), [Pinto Bea...",$8.49


8.  The 'item\_price' column is currently a string (e.g., '$2.39'). Convert it to a float type. You'll need to remove the dollar sign and then convert the data type.


Remove the dollar sign
Convert it to a float type.

In [54]:
df["item_price"] = df["item_price"].str.lstrip('$')
df["item_price"] = df["item_price"].astype(float)
df.head()

,order_id,quantity,item_name,choice_description,item_price
0,1,1,Chips and Fresh Tomato Salsa,NaN,2.39
1,1,1,Izze,[Clementine],3.39
2,1,1,Nantucket Nectar,[Apple],3.39
3,1,1,Chips and Tomatillo-Green Chili Salsa,NaN,2.39
4,2,2,Chicken Bowl,"[Tomatillo-Red Chili Salsa (Hot), [Black Beans...",16.98


9.  After converting 'item\_price' to a numerical type, calculate the average `item_price` across all orders.


average price acrosss all orders


In [55]:
df.item_price.mean()

np.float64(7.464335785374297)

Count no of unique orders

10. Determine the total number of unique orders in the dataset based on the `order_id` column.

In [56]:
df3 = df.drop_duplicates(subset=['order_id'])
df3.order_id.count()

np.int64(1834)

Section 3: Optional; Extra Credit (2 Questions)

Total Revenue Calculation (Pandas)

Assuming each item_price (after conversion to float) is for a single quantity of that item, calculate the total revenue generated from all orders. Remember that the quantity column indicates how many of that item_name were ordered in that specific line item.

How calculate grand total

In [57]:
df["grand_total"] = df["quantity"] * df["item_price"]
df.head()

,order_id,quantity,item_name,choice_description,item_price,grand_total
0,1,1,Chips and Fresh Tomato Salsa,NaN,2.39,2.39
1,1,1,Izze,[Clementine],3.39,3.39
2,1,1,Nantucket Nectar,[Apple],3.39,3.39
3,1,1,Chips and Tomatillo-Green Chili Salsa,NaN,2.39,2.39
4,2,2,Chicken Bowl,"[Tomatillo-Red Chili Salsa (Hot), [Black Beans...",16.98,33.96


Calculate Total Sales per Item (PySpark):

Convert the pandas Chipotle dataframe to a PySpark dataframe. Transform the 'item_price' column to a numeric type, and calculate the line_item_total (quantity * item_price) for each row. Then, calculate the total sales for each unique item_name and display the top 5 items by their total sales.

Using Pyspark to calculate total per item
requires one to know how to group by using pyspark

First convert df to Pyspark df

In [58]:
from pyspark.sql import SparkSession
import pandas as pd

spark = SparkSession.builder.appName("example").getOrCreate()
df3  = pd.read_csv("https://raw.githubusercontent.com/justmarkham/DAT8/master/data/chipotle.tsv", sep='\t')
spark_df = spark.createDataFrame(df3)
spark_df.show()

+--------+--------+--------------------+--------------------+----------+
|order_id|quantity|           item_name|  choice_description|item_price|
+--------+--------+--------------------+--------------------+----------+
|       1|       1|Chips and Fresh T...|                 NaN|    $2.39 |
|       1|       1|                Izze|        [Clementine]|    $3.39 |
|       1|       1|    Nantucket Nectar|             [Apple]|    $3.39 |
|       1|       1|Chips and Tomatil...|                 NaN|    $2.39 |
|       2|       2|        Chicken Bowl|[Tomatillo-Red Ch...|   $16.98 |
|       3|       1|        Chicken Bowl|[Fresh Tomato Sal...|   $10.98 |
|       3|       1|       Side of Chips|                 NaN|    $1.69 |
|       4|       1|       Steak Burrito|[Tomatillo Red Ch...|   $11.75 |
|       4|       1|    Steak Soft Tacos|[Tomatillo Green ...|    $9.25 |
|       5|       1|       Steak Burrito|[Fresh Tomato Sal...|    $9.25 |
|       5|       1| Chips and Guacamole|           

Now convert item_price to numeric


In [59]:
pd = spark_df.toPandas()
pd["item_price"] = pd.item_price.str.lstrip('$')
pd.head()
spark_df = spark.createDataFrame(pd)
spark_df.show()

+--------+--------+--------------------+--------------------+----------+
|order_id|quantity|           item_name|  choice_description|item_price|
+--------+--------+--------------------+--------------------+----------+
|       1|       1|Chips and Fresh T...|                 NaN|     2.39 |
|       1|       1|                Izze|        [Clementine]|     3.39 |
|       1|       1|    Nantucket Nectar|             [Apple]|     3.39 |
|       1|       1|Chips and Tomatil...|                 NaN|     2.39 |
|       2|       2|        Chicken Bowl|[Tomatillo-Red Ch...|    16.98 |
|       3|       1|        Chicken Bowl|[Fresh Tomato Sal...|    10.98 |
|       3|       1|       Side of Chips|                 NaN|     1.69 |
|       4|       1|       Steak Burrito|[Tomatillo Red Ch...|    11.75 |
|       4|       1|    Steak Soft Tacos|[Tomatillo Green ...|     9.25 |
|       5|       1|       Steak Burrito|[Fresh Tomato Sal...|     9.25 |
|       5|       1| Chips and Guacamole|           

In [61]:
from pyspark.sql.functions import col
spark_df= spark_df.withColumn("grand_total", col("item_price") * col("item_price")).groupBy("item_name").sum("grand_total")
spark_df.show()


AnalysisException: [UNRESOLVED_COLUMN.WITH_SUGGESTION] A column or function parameter with name `item_price` cannot be resolved. Did you mean one of the following? [`item_name`, `sum(grand_total)`].;
'Project [item_name#260, sum(grand_total)#303, ('item_price * 'item_price) AS grand_total#318]
+- Aggregate [item_name#260], [item_name#260, sum(grand_total#289) AS sum(grand_total)#303]
   +- Project [order_id#258L, quantity#259L, item_name#260, choice_description#261, item_price#262, (cast(item_price#262 as double) * cast(item_price#262 as double)) AS grand_total#289]
      +- LogicalRDD [order_id#258L, quantity#259L, item_name#260, choice_description#261, item_price#262], false
